tracing-subscriber/src/registry/sharded.rs

```diff
@@ -0,0 +1,242 @@
+use sharded_slab::{Guard, Slab};
+
+use crate::{
+    fmt::span::SpanStack,
+    registry::{LookupSpan, SpanData},
+};
+use std::{
+    cell::RefCell,
+    collections::HashMap,
+    convert::TryInto,
+    fmt,
+    sync::{
+        atomic::{fence, AtomicUsize, Ordering},
+        Mutex,
+    },
+};
+use tracing_core::{
+    dispatcher,
+    field::{FieldSet, Visit},
+    span::{self, Id},
+    Event, Field, Interest, Metadata, Subscriber,
+};
+
+#[derive(Debug)]
+pub struct Registry {
+    spans: Slab<Data>,
+}
+
+thread_local! {
+    static CONTEXT: RefCell<SpanStack> = RefCell::new(SpanStack::new());
+}
+
+#[derive(Debug)]
+pub struct Data {
+    metadata: &'static Metadata<'static>,
+    parent: Option<Id>,
+    children: Vec<Id>,
+    ref_count: AtomicUsize,
+}
+
+// === impl Registry ===
+
+impl Default for Registry {
+    fn default() -> Self {
+        Self { spans: Slab::new() }
+    }
+}
+
+#[inline]
+fn idx_to_id(idx: usize) -> Id {
+    Id::from_u64(idx as u64 + 1)
+}
+
+#[inline]
+fn id_to_idx(id: &Id) -> usize {
+    id.into_u64() as usize - 1
+}
+
+impl Registry {
+    fn insert(&self, s: Data) -> Option<usize> {
+        self.spans.insert(s)
+    }
+
+    fn get(&self, id: &Id) -> Option<Guard<'_, Data>> {
+        self.spans.get(id_to_idx(id))
+    }
+}
+
+thread_local! {
+    static CURRENT_SPANS: RefCell<SpanStack> = RefCell::new(SpanStack::new());
+}
+
+impl Subscriber for Registry {
+    fn register_callsite(&self, _: &'static Metadata<'static>) -> Interest {
+        Interest::always()
+    }
+
+    fn enabled(&self, _: &Metadata<'_>) -> bool {
+        true
+    }
+
+    #[inline]
+    fn new_span(&self, attrs: &span::Attributes<'_>) -> span::Id {
+        let parent = attrs.parent().map(|id| self.clone_span(id));
+        let s = Data {
+            metadata: attrs.metadata(),
+            parent,
+            children: vec![],
+            ref_count: AtomicUsize::new(1),
+        };
+        let id = self.insert(s).expect("Unable to allocate another span");
+        idx_to_id(id)
+    }
+
+    #[inline]
+    fn record(&self, _: &span::Id, _: &span::Record<'_>) {
+        // TODO(david)—implement this.
```

Comment:
actually, i don't think we _should_ implement this; layers will do that part. :)

Reply:
That makes sense, but `Subscriber::record` is a required method on `Subscriber`, unfortunately. This might might require changing `Subscriber` to support this, I think.

Reply:
i just meant that it should remain an empty no-op method...

Reply:
oh! okay, yeah. makes sense!